### ISSUE TITLE
Using GOAP system for jobs

### ISSUE THREAD

**Issue by NogginBops** [TeamPorcupine/ProjectPorcupine#86]
So this is what we came up with after a long discussion in #75 and #63.

The plan for the new job system and really the new AI system is to use GOAP (Goal Oriented Action Planning).

I won't go into the finer workings of GOAP but there are some great resources out there:
[A GOAP rescource website](http://alumni.media.mit.edu/~jorkin/goap.html)
[TheHappieCats video](https://www.youtube.com/watch?v=n6vn7d5R_2c)
[Quill18s video](https://www.youtube.com/watch?v=bc86es4YOoc)
[A paper on applying GOAP in practice](http://alumni.media.mit.edu/~jorkin/GOAP_draft_AIWisdom2_2003.pdf)

The general gist is to not worry about dependencies and special/edge cases and only worry about the end goal.

So all the programmer has to do is just specify the desired end result and the system will figure out what needs to be done for the goal to be reached.

This is done by defining actions that a character can take. A action contains what dependencies that action takes (e.g Building a wall requires building materials) and what outcome that action results in (e.g the wall being built). In addition to this a action contains a cost function that is used to calculate what the cost of a action is. This can be in either in respects to resources or time (Efficiency vs. time).

So when a goal is specified it goes through all the available actions that result in the desired result and chooses the least costly solution.

This is a good solution because it allows modders to easily create cross mod support. For example consider the following scenario:

Modder x creates a mod that creates a machine that creates steel plates super efficiently. 
And modder y creates a mod that requires steel plates to craft a really cool weapon. 

Using some kind of other system, a specific case for the other mods machine might be needed. 
But with GOAP this is a non issue. The machine in modder xs mod would just specify an action to create steel plates and modder ys mod would specify a goal that requires steel pates. The system would then figure out that crafting the steel plates is the cheapest way to create steel plates and create the steel plates that way. 
And just like that, Done!

---

Goals that should be solved (please add more):
- Ensure that close jobs are picked first, unless there is something urgent, so that what the character does makes sense to the player.

**Comment by lameox** [comment]
I think we should keep this Proposal up for a few days before starting the implementation so everybody has a chance to look at it, give some critizism, and participate.

**Comment by NogginBops** [comment]
@lameox Totally agree,

**Comment by vogonistic** [comment]
Agreed. Should also start fleshing out what the implementation of a Porcupine GOAP would look like.

**Comment by lameox** [comment]
I think we could do something like Natural Needs (Drinking, Sleeping) of our characters and also Artificial Needs (Need fuel for generators) of the whole station which would generate a ranking of desired goals. we could then have a weighing function for each need i.e. drinking is exponentially more urgent the thirstier our character gets. Speaking of that, we should probably design the system in a way, that an urge can get fulfilled before it is too late (character dies of thirst).

As i said previously i havent really worked with GOAP before, but it seems to me like those needs that generate goals should be like the first layer in our decision algorithm and then the GOAP takes the goal and calculates an action plan.

Thoughts?

**Comment by emilkris33** [comment]
@lameox Sounds about right. We definitely need Needs and a system where drinking when really thirsty or say firefighting happens before building a new wall. And while a priority system needs to be thought into the GOAP i think we can wait with needs til are GOAP works and other basic functionality is in.

And by the way, it looks like the back and forth discussion has moved the job system idea to a really nice place.

**Comment by lameox** [comment]
Yeah the discussion really moved us in the right direction.

I think we should think about the interface though. For now we can stub it out but I think it would be advantageous to have an idea of our interface in mind before implementing the GOAP

**Comment by NogginBops** [comment]
@lameox I agree with you that we need a system that priorities needs and such but I also agree with @emilkris33 that the underlying GOAP system implementation should be prioritized first.

And I hope so @emilkris33, we have been going at it for a few hours. :P (I like the idea we are heading with too! :D Think it's going to work really well)

**Comment by NogginBops** [comment]
@lameox By interface do you mean the UI or the coding interface (api)?

**Comment by lameox** [comment]
@NogginBops Coding interface. I had thought each Manifestation of a Need would implement some interface against which we can implement the goals. in doing so we get an easily testable System because we can provide stubbed implementations of that interface when testing.

**Comment by NogginBops** [comment]
@lameox I don't fully understand what you are suggesting. Care to elaborate?

**Comment by lameox** [comment]
We define an interface that our Goals will implement, i.e. 
```
public interface IGoal {
    List<Item> Requires  { get; }
    List<Item> Produces { get; }
    List<Condition> Proconditions { get; }
}

class Goal : IGoal {
 //implementation
}
```

This will be our object that we use in the GOAP. For example we may need stuff like the Requires, Produces and Procodition fields. This will be the interface we program against. This means that in our GOAP implementation we will only ever call Methods and access fields defined in IGoal. Then when we want to write unit tests we can provide a dummy implementation of that interface and use it in our tests. by doing so we can easily test specific parts of the code without having to setup a huge net of dependencies we would have in a normal run.

If you want to read up on it:
[MSDN Interfaces](https://msdn.microsoft.com/en-us/library/ms173156.aspx)
[Unit testing](https://msdn.microsoft.com/en-us/library/hh694602.aspx)
[Stubbing](https://msdn.microsoft.com/en-us/library/hh549174.aspx)

**Comment by NogginBops** [comment]
I have read up and thought about this issue for some time now and I think I'm ready to start implementing this. Though I want the conversation up and running again.

The idea of using GOAP only provides a general description of the actual algorithm used, but nothing about the actual code implementation. So that is what I what to ask.

The algorithm requires some kind of understanding of the world, this is usually implemented via a FSM (Finite State Machine) of some kind. And the actions usually kind of represent the transition between these states. Is this how we want the system to operate or is there a better solution more fitting for our problem?

Another thing is that to be able to create a valid action tree the ability to compare and link prerequisites to results and vise versa becomes important. How do we want to represent the prerequisites and result in data? How generalized should we make this? (Use this system for all is or just for the job system)

I think this is a good starting off point to get this conversation going. Though there are many more things to consider.

@lamexo @emilkris33 @vogonistic

**Comment by emilkris33** [comment]
@NogginBops I think this is such a big system that i needs a fairly detailed plan before any work on it can start. So we likely need someone to write up some sort of detailed description of how we want the system to work before coding starts. Also for such a big system we should have a word from @quill18 before starting implementation.

**Comment by vogonistic** [comment]
Maybe it's worth converting the Character to be based on a FSM, but that almost feels like a separate PR in itself. Remember that the more you touch in this turbulent repo, the more of other peoples changes you'll have to merge. I'm sure it could be implemented into the current system as well.

If you are volunteering to main dev on this, I say you should follow the principle that Quill has been using in the tutorial. Start building a small thing that works and then go back and improve it and add more edge cases and such.

If you create a branch on your repo, I can add it as a remote and help you test, look at code and make some PR's against it. I'm not sure what you level of git/github knowledge is, so let me know if you want someone else to do it instead. I think it would be beneficial to have a feature branch that is worked against for this though.

As far as scope goes, start with something simple-ish like solving the reasoning for BuildWall -> FindSteel -> WorkMiningDroid.

Have you found any information on how sticky the reasoning is? Is it regenerated every turn or cached until it is cancelled?

As far as resolving the paths, you can implement it like an A* where you have a start point and a goal condition and a list of functions that can resolve things. Each pass you pop an intermediary condition, loop over the resolver functions and ask if they can provide something that converts that. You check if it fulfilled the goal, gave you another intermediary (or several) or just failed to do anything with it. The hard part is to figure out of the conversion is going the right way or not, the best cost would be time. Walking takes time based on distance, operating the drone also takes time.

**Comment by vogonistic** [comment]
@emilkris33 I get the feeling that Quill is seeing where people are taking it rather than setting a fixed path. I'm pretty sure he'd read this and tell us if we are completely off the path he wants to go.

Either way, I'm excited to try building a GOAP.

**Comment by lameox** [comment]
I think having some kind of global representation of our state would be a good way to start. This way we can implement all actions in a way that just describes how the state will be changed.

**Comment by NogginBops** [comment]
@vogonistic Yeah, I would be happy to main dev this. 
I agree with @emilkris33 that there should be a pretty detailed plan before just blindly trying to implement this. I also agree with @lameox that we should probably hold of on doing any of the real work before things settle, though starting to flesh out a plan of action is not a bad idea. I wanted for the implementation details conversation to get a head start before we do any of the real work.

@lameox I agree with you that a FSM is a good place to start. But what is the data we care to model? What states should exist?

@vogonistic The A* for resolving paths is a good idea. A good version of A* for this job is [IDA*](https://en.wikipedia.org/wiki/Iterative_deepening_A*)(Iterative deepening A*) because of efficient memory usage and easy depth limit.

I want to get feedback on the idea of using this system as a system for all ai in the game and not just for jobs. I mean GOAP is already designed to work for any intelligent actor.

**Comment by lameox** [comment]
I think the states we need should be a representation of positions, items, and object/character states (think ill, injured, smalting steel). I think those 3 types of states should allow us to capture enough information about the world to model most actions for now. we might later introduce aditional states if the need arises.

**Comment by NogginBops** [comment]
@lameox Is a FSM really needed to model that data though? Dosen't World.cs contain all that information, if yes then why would we need to create a FSM we could just define a set of procedures for changing the state of the world? What are your thoughts?

**Comment by Kreol-Dev** [comment]
Wouldn't HTN + Utility functions be more effective? From my experience resource managing GOAP is a mess if you make it anything but "specific action always requires same amount of resources". Any interdependency between them (like, fuel per km when car is moving) gonna break the whole idea of simple AI that can account for that, because actions will become bloated with accounting for different cases.

GOAP is great for tactical combat, not so much in this case.

**Comment by NogginBops** [comment]
@Kreol-Dev With HTN do you mean Hierarchical task network?

**Comment by NogginBops** [comment]
I have read the Wikipedia article and this seems like it could be the system I hoped for but wasn't able to come up with and didn't quite know about existed. That's what happens when you don't have a lot of experience around the field.

**Comment by Kreol-Dev** [comment]
@NogginBops Yep, 

Because, for example, a case of a metal producing machine can be solved in much simpler way just by creating a "production component", so actions won't really orient for specific furniture, but for an interfacing component. 
If it will require resource, then you would create a "consumer" component, and a Task that will search for both producer and consumer on the same object and so one.

**Comment by NogginBops** [comment]
I will read up on this topic to understand more of the different concepts but it seems promising.

Though I want some feedback from @lameox @emilkris33 @vogonistic that have been active in this thread, about what they think about this?

**Comment by NogginBops** [comment]
@Kreol-Dev Yeah, your case makes sense.

**Comment by WardBenjamin** [comment]
Incidentally, both Prison Architect and Rimworld handle jobs in pretty much this system and it seems to work well for both. I definitely like the get materials -> stockpile them in-place -> construction paradigm, but it probably should be paired with a priority queue, since it doesn't really matter *who* (as in, which character) does each task, and some tasks should be higher priority than others (that should probably be set by the user, but we should also have a default). 

This priority system could also be used to enqueue tasks in order, just make tasks that others depend on higher priority and then sort. That would filter out tasks based on both user-set priority and would put them in the order required for production chains.

**Comment by NogginBops** [comment]
@WardBenjamin What are your thoughts on a [HSN](https://en.wikipedia.org/wiki/Hierarchical_task_network#cite_note-6) implementation instead of a GOAP one?

**Comment by vogonistic** [comment]
The implementation in my head is probably leaning towards HSN, but I'm good either way.

**Comment by WardBenjamin** [comment]
Agreed, HSN seems like a better solution.

**Comment by koosemose** [comment]
I'll be honest, I'm only able to follow the rough idea of what you all are planning on, but I do have a question. Is this system intended to only cover how to perform a job or would it also cover what jobs are being done?

**Comment by vogonistic** [comment]
@koosemose It would have to at the very least tie into the system that prioritizes what jobs are done first, but it's not the same thing. Do you want to start a separate discussion on that?

**Comment by kd7uiy** [comment]
I submitted a very simple change proposal to add a priority queue system that works and is a huge improvement, see https://github.com/TeamPorcupine/ProjectPorcupine/pull/293

**Comment by NogginBops** [comment]
@koosemose I think the idea right now is to have this system incorporate all AI in the game. This would both affect how the jobs are done and as an extension what/why jobs are being done.

**Comment by koosemose** [comment]
@vogonistic  No, I was just trying to understand exactly what was being accomplished. And I was worried that it was intended to handle the actual prioritising of jobs ( as it would perhaps take away too much player choice ). As is, it looks good, and I just request that you all keep in mind that partially a tutorial, and to endeavor to ensure that this complex sounding system can be presented in such a way as to not scare off novice programmers from any parts of the code it touches.

**Comment by NogginBops** [comment]
@koosemose That is absolutely a important thing to consider and a part of the solution I think lays in a clear and well defined structure and clear api for which this thread exists.

**Comment by WardBenjamin** [comment]
> One possible solution to this is to pair it with an HTN, so that you would have a priority queue for each Goal (a meta-job, filled with smaller jobs that depend on one another), and then another queue internal to each Goal to order the jobs. If the Goal becomes gridlocked, you could either bump it down in the Goal queue or notify the end-user, or both.

(Discussion from @kd7uiy's partial job fix)

**Comment by lameox** [comment]
I will need to read up on both GOAP and HTN more before i feel like i can give any good feedback on the matter. I will have to see about finding time for that though...

**Comment by Kreol-Dev** [comment]
I would highly suggest such structure for HTN AI:

Create abstract "Task" class that all tasks would derive. Starting from "TaskType" which adds context, which means specific data the task needs to know to be performed.
For each task type (for example, AcquireItemsInPlace(Item, Place)) you create the subclass of Task,

Each speficic primitive or compound tasks would then derive from task types. 

Each task, apart from "perform" functions\coroutine should have a utility function and "object context", that means "What kind of object can perform this task". That way in future, for example, an agent without hands component couldn't perform tasks like "PushButton".

If an agent can perform more then one task of certain type (for example, it can move items by hand, or teleport them using magick), what specific task to choose is determined by it's utility.

At the same time, the main task is also choosen by utility, but a bit more complex.

Any "colony wide need" - like "having 10 metal" - should be placed in a blackboard that holds all the needs. 

When an agent would need to perfrom a task, apart from own needs (like food), tasks could refer to "colony needs" to determine their utility.

That way we could also introduce fun mechanics, that I haven't seen in such games - altruism vs egoism in agents. So, an egoistic agent will prefer satisfying it's own needs or "dangerous to him" colony needs, like defence, while altruistic might forget about self and value "colony well-being" more.

**Comment by Mizipzor** [comment]
Should we get the ball rolling on this? For this game, this system is probably by far the most important one. I see no problem with multiple approaches being worked on by different people as we investigate. Can I help? :)

**Comment by kd7uiy** [comment]
I've got a very primitive version of this going, it allows each job to set a priority, which adds it to a queue. It should be approved soon, I had to merge it, but Quill otherwise liked it. I fixed the build, and hopefully he'll pull it in to the master build soon.

**Comment by vogonistic** [comment]
This is going to be a habit of mine, I'm sure, but I'm going to quote Quill from the Gatekeeper post #321:
> I'd rather do something slightly wrong now than debate for weeks over the perfect choice.

This system, regardless of GOAP vs HTN, will be very complicated. Does someone want to build a small model of it decoupled from the game? A couple of classes that does the right thing in an ideal circumstance without integration to World? We could have people build a couple of these, or collaborate on them. I'm sure a few of you have started something. =)

**Comment by kd7uiy** [comment]
The best way to do any such task is to break it in to parts. That's part of why I pushed my primitive improvement, it's better than what's there, and it is starting the conversation to come up with the right questions to ask.

**Comment by vogonistic** [comment]
@Kreol-Dev What version of Unity are you using? The system sounds pretty awesome.

@kd7uiy What do you think should be done first?

@NogginBops Which direction are you leaning now with the implementation? Started working on something?

**Comment by vogonistic** [comment]
I added a goal to the bottom of the top post in this issue, to capture issue #308 by @itay347.

We should probably expand the goals further.

**Comment by Kreol-Dev** [comment]
@vogonistic 5.3, as with my other projects

**Comment by WardBenjamin** [comment]
@Kreol-Dev FYI, Unity 5.3 doesn't actually work with this project due to using Random.InitState() and some prefab differences.

**Comment by Kreol-Dev** [comment]
@WardBenjamin then I can only help with advice and blessing :D

**Comment by Mizipzor** [comment]
Is there an issue for that initstate thing?

**Comment by WardBenjamin** [comment]
@Kreol-Dev why don't you update to 5.4? You can even have multiple versions installed if you want. 

@Mizipzor It's not really an issue because our prefabs also depend on 5.4, so there's no reason to fix it. Plus it's useful :P

**Comment by Kreol-Dev** [comment]
@WardBenjamin PP is for 5.4? I thought it for earlier version and didn't want to have backward version of Unity installed.
That way I now have one more reason to upgrade. Propably gonna do that next morning.

**Comment by gunthergun** [comment]
Is somebody working on this system at the moment ?

**Comment by NogginBops** [comment]
@gunrhergun I don't thinks so. Note that we kindof decided on a HTN  rather than GOAP.

**Comment by vogonistic** [comment]
I have the start of some code I was playing with before I got picked as a gatekeeper. It can sort of do simple optimizations, but it's pretty naive code that isn't fully thought out. If anyone is interested, I could share it. It's not connected to the actual job queue, but could be used as inspiration.

**Comment by vogonistic** [comment]
It's early code so take it for what it is: https://github.com/vogonistic/ProjectPorcupine/commit/354bd8124fc41299a6f23224121512cae189e336

It's on a branch called `job-optimizer` in my repo. It predates @Kreol-Dev, but I was going for something kinda like that.

**Comment by vogonistic** [comment]
@gunthergun Are you working on this?

**Comment by gunthergun** [comment]
currently not no, just wanted to know how things were going

**Comment by vogonistic** [comment]
I'm closing this thread. We can continue discussing in #726 and other threads.
